src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -169,30 +179,31 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
 
         private static async Task<Socket> ConnectAsync(string serverName, int port)
         {
+            Socket socket;
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 await socket.ConnectAsync(serverName, port).ConfigureAwait(false);
                 return socket;
             }
 
-            // On unix we can't use the instance Socket methods that take multiple endpoints
-
+            // On unix we can only connect using IP addresses, rather than host names, using the instance connect methods.
             IPAddress[] addresses = await Dns.GetHostAddressesAsync(serverName).ConfigureAwait(false);
-            return await ConnectAsync(addresses, port).ConfigureAwait(false);
-        }
-
-        private static async Task<Socket> ConnectAsync(IPAddress[] serverAddresses, int port)
-        {
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            if (addresses == null || addresses.Length == 0)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
-                await socket.ConnectAsync(serverAddresses, port).ConfigureAwait(false);
-                return socket;
+                // Connection error will be thrown by the caller after receiving a null socket.
+                return null;
             }
 
-            // On unix we can't use the instance Socket methods that take multiple endpoints
+            IPAddress targetAddr = addresses[0];
```

Comment:
Changed the logic here to single IP, since this method only gets called if parallel connect is disabled.